ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InMemoryKVParticipantState.scala

```diff
@@ -0,0 +1,216 @@
+import com.digitalasset.platform.akkastreams.Dispatcher
+import com.digitalasset.platform.akkastreams.SteppingMode.OneAfterAnother
+import com.digitalasset.platform.services.time.TimeModel
+import com.google.protobuf.ByteString
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+import scala.collection.breakOut
+
+object InMemoryKVParticipantState {
+
+  /** The complete state of the participant at a given point in time.
+    * This emulates a key-value blockchain with a log of commits and a key-value store.
+    * The commit log provides the ordering for the log entries, and its height is used
+    * as the [[Offset]].
+    * */
+  case class State(
+      log: Vector[(DamlLogEntryId, DamlSubmission)],
+      recordTime: Timestamp,
+      // Store is the key-value store that the commits mutate.
+      store: Map[ByteString, ByteString],
+      config: Configuration
+  )
+
+  // Message sent to the commit actor to commit a submission to the state.
+  case class CommitMessage(
+      // Submitter chosen entry identifier
+      entryId: DamlLogEntryId,
+      submission: DamlSubmission
+  )
+}
+
+/* Implementation of the participant-state using the key-value utilities and an in-memory map. */
+class InMemoryKVParticipantState(implicit system: ActorSystem)
+    extends ReadService
+    with WriteService
+    with AutoCloseable {
+  import InMemoryKVParticipantState._
+
+  val ledgerId = SimpleString.assertFromString(UUID.randomUUID.toString)
+
+  // DAML Engine for transaction validation.
+  private val engine = Engine()
+
+  // Random number generator for generating unique entry identifiers.
+  private val rng = new java.util.Random
+
+  // Namespace prefix for log entries.
+  private val NS_LOG_ENTRIES = ByteString.copyFromUtf8("L")
+
+  // Namespace prefix for DAML state.
+  private val NS_DAML_STATE = ByteString.copyFromUtf8("DS")
+
+  // Reference to the latest immutable state. Reference is only updated by the CommitActor.
+  // Reading from the state must happen by first taking the reference (val state = stateRef).
+  @volatile private var stateRef: State =
+    State(
+      log = Vector.empty[(DamlLogEntryId, DamlSubmission)],
+      recordTime = Timestamp.Epoch,
+      store = Map.empty[ByteString, ByteString],
+      config = Configuration(
+        timeModel = TimeModel.reasonableDefault
+      )
+    )
+
+  // Akka actor that receives submissions and commits them to the key-value store.
+  @SuppressWarnings(Array("org.wartremover.warts.Any"))
+  class CommitActor extends Actor {
+    override def receive: Receive = {
+      case CommitMessage(entryId, submission) =>
+        val state = stateRef
+        if (state.store.contains(entryId.getEntryId)) {
+          // The entry identifier already in use, drop the message and let the
+          // client retry submission.
+        } else {
+          // Process the submission to produce the log entry and the state updates.
```

Comment:
add appropriate debug log messages that allow tracing how validation etc. progresses